apps/slack/src/lib/WebhookActivityToggler.service.ts

```diff
@@ -0,0 +1,119 @@
+import { Client, OperationResult } from "urql";
+import {
+  DisableWebhookDocument,
+  EnableWebhookDocument,
+  FetchOwnWebhooksDocument,
+  WebhookEventTypeEnum,
+} from "../../generated/graphql";
+
+export interface IWebhookActivityTogglerService {
+  disableOwnWebhooks(webhooksIdsParam?: string[]): Promise<void>;
+  enableOwnWebhooks(): Promise<void>;
+}
+
+/**
+ * Original implementation in apps/search.
+ *
+ * TODO: Create common abstraction, shared by all apps.
+ */
+export class WebhooksActivityClient {
+  constructor(private client: Pick<Client, "query" | "mutation">) {}
+
+  private handleOperationFailure(r: OperationResult) {
+    if (r.error || !r.data) {
+      console.error(
+        {
+          error: r.error,
+        },
+        "Error disabling webhook",
+      );
+      throw new Error("Error disabling webhook");
+    }
+  }
+
+  fetchAppWebhooksIDs(id: string) {
+    return this.client
+      .query(FetchOwnWebhooksDocument, { id })
+      .toPromise()
+      .then((r) => {
+        this.handleOperationFailure(r);
+
+        if (!r.data?.app?.webhooks) {
+          throw new Error("Webhooks not registered for app, something is wrong");
+        }
+
+        return r.data?.app?.webhooks?.map((w) => w.id);
+      });
+  }
+
+  disableSingleWebhook(id: string): Promise<void> {
+    return this.client
+      .mutation(DisableWebhookDocument, {
+        id,
+      })
+      .toPromise()
+      .then((r) => {
+        this.handleOperationFailure(r);
+
+        return undefined;
+      });
+  }
+
+  enableSingleWebhook(id: string): Promise<void> {
+    return this.client
+      .mutation(EnableWebhookDocument, {
+        id,
+      })
+      .toPromise()
+      .then((r) => {
+        this.handleOperationFailure(r);
+
+        return undefined;
+      });
+  }
+}
+
+export class WebhookActivityTogglerService implements IWebhookActivityTogglerService {
+  /**
+   * Extracted separate client for easier testing without touching graphQL
+   */
+  private webhooksClient: WebhooksActivityClient;
+
+  constructor(
+    private ownAppId: string,
+    private client: Pick<Client, "query" | "mutation">,
+    options?: {
+      WebhooksClient: WebhooksActivityClient;
+    },
+  ) {
+    this.webhooksClient = options?.WebhooksClient ?? new WebhooksActivityClient(this.client);
+  }
```

Comment:
Why do we construct `WebhooksActivityClient` in here? Even if you want to use the service with a mocked `options.WebhookClient` you still need to provide a mock for `client`, even though it's not used anywhere in the service itself - only for initialization

Reply:
Its dependency injection. If you don't provide options, it will automatically create instance. But for testing, you can create any mocked instance of `WebhooksActivityClient`.

I agree that this implementation should be improved, so graphql client is not needed. It should be hidden behind injected webhooksClient. But this is implementation copy-pasted from Search app.

I'm happy to fix this, but this should probably happen with extraction of this service to shared package and later implementing to every other app. 

This PR is meant to quickly implement performance optimization, I will add ticket for it